packages/nimble-components/src/rich-text/specs/mention-hld.md

```diff
@@ -0,0 +1,650 @@
+    pattern in the configuration element, even the absolute links will transform into `mention` nodes.
+
+#### Alternatives
+
+Slot based approach: All user information will be communicated through `nimble-list-option` elements. These elements will
+contain the user's ID in their `value` attribute and the user's name in their text content. When the **"@"** symbol is
+inserted into the editor, these elements will be presented as a dropdown list.
+
+Pros:
+
+1. When the client does not specify a `value` attribute, the options are sent without a user's ID, and in such cases,
+   the value attribute is derived from the content of the option.
+
+Cons:
+
+1. It is impossible to have multiple type of mentions when `nimble-list-option` is used in the default slot. If we use
+   slot names to differentiate between various type of mentions, the rich text editor needs to be aware of all the type
+   of mentions like user slot name, issue slot name etc, through an attribute value.
+2. The slot-based approach is not configurable, meaning that if a user wishes to customize the entire user mention with
+   some arbitrary value, the only viable option is to include it within the options. In contrast, the configuration-based
+   approach readily allows for the inclusion of arbitrary values in the parent container element, making customization more
+   straightforward. For an example, in future if we are configuring the `character` to trigger the user mention from the client,
+   it is simply to add it like below,
+    ```html
+    <nimble-rich-text-mention-users character="@">
+        <nimble-mapping-mention-users
+            key="user-id-1"
+            text="username-1"
+        ></nimble-mapping-mention-users>
+        <nimble-mapping-mention-users
+            key="user-id-2"
+            text="username-2"
+        ></nimble-mapping-mention-users>
+        <nimble-mapping-mention-users
+            key="user-id-3"
+            text="username-3"
+        ></nimble-mapping-mention-users>
+    </nimble-rich-text-mention-users>
+    ```
+    If it is a slot based approach, it is required to provide the information to every options as we don't know that it is a user mention
+    or other mention
+    ```html
+    <nimble-mapping-mention-users
+        key="user-id-1"
+        text="username-1"
+        character="@"
+    ></nimble-mapping-mention-users>
+    <nimble-mapping-mention-users
+        key="user-id-2"
+        text="username-2"
+        character="@"
+    ></nimble-mapping-mention-users>
+    <nimble-mapping-mention-users
+        key="user-id-3"
+        text="username-3"
+        character="@"
+    ></nimble-mapping-mention-users>
+    ```
+3. The UI of the list option is completely dependent on the client application.
+4. All the keyboard and mouse interactions within the dropdown list should be handled internally within the editor as `nimble-list-option`
+   does not support any.
+
+#### Client Usage Guidance on Filtered Users:
```

Comment:
@rajsite and I brainstormed another case to cover. The list of user mappings provided to the rich text editor will need to include a union of:
    1. the users that match the current @mention substring (as you've described already)
    2. all other users that are mentioned in the editor
Item 2 is necessary so that those users can render with their display name rather than their ID.

This could complicate the client logic a bit. The client will need to be querying the editor for its list of mentioned users to populate 2.

It also opens UX questions. If some users are mentioned already then the mapping list will never be empty. So if the user types `@` then the dropdown would already be populated with the other mentioned users. This seems fairly useful since it seems probable that someone may want to mention someone else multiple times. As the user types some filter after the `@`, the previously mentioned users that don't match the filter will be removed by the rich text editor's filter.

Reply:
Yes, we should keep the state of both the filtered option and already mentioned users to let the editor know how to render the elements. I just updated with a few more details on this section in the HLD to handle the above case.

As I mentioned in the HLD, the client can maintain two lists, one stores the already mentioned users which will stay as children of the editor for the whole time, and the other one is dynamic filtering mapping elements that will change based on the event the config element emits for filtering.

As you said, the already-mentioned users will be there in the popup like the recently mentioned users and filtered out when they type by the editor's filtering logic. We think this could be the expected behavior and can be doable in the way I mentioned in the doc for the clients.